Separate Params / Metrics

pearsonca opened the pull request:
finally getting to the capability I want to add to AbcSmc: otherwise-specifiable-pseudo params.

This is currently a draft, but can at the new headers / cpps to get a feel for how it works.

tjhladish commented on examples/pseudo/config.json:
```diff
@@ -0,0 +1,34 @@
+         "dist_type"  : "PSEUDO",
+         "num_type"   : "INT",
+         "par1"       : 1,
+         "par2"       : 5
```
It would be nice to show in at least one example that step is an available argument

pearsonca replied:
plan to resolve this via #46 discussion

tjhladish commented on include/AbcSmc/AbcSmc.h:
```diff
@@ -4,298 +4,110 @@
 
-        FilteringType get_filtering_type() const { return use_pls_filtering ? PLS_FILTERING : SIMPLE_FILTERING; }
+        // should be private?
+        void set_filtering_type(const ABC::FILTER & ft) { _filtering = ft; }
```
Remove spaces following & (here and elsewhere)

pearsonca replied:
...several of these, but pretty sure all caught.

tjhladish commented on include/AbcSmc/IndexedPars.h:
```diff
@@ -0,0 +1,59 @@
+
+    // TODO support other constructors?
+
+    float_type sample(PRNG & prng) const override { return states[prng.pseudo(this)]; }
```
maybe be consistent with prng vs rng name

tjhladish commented on include/AbcSmc/Metric.h:
```diff
@@ -0,0 +1,37 @@
+            std::string short_name;
+    };
+
+    template <typename NT>
```
Add comment explaining TMetric is a typed metric

tjhladish commented on include/AbcSmc/ParXform.h:
```diff
@@ -0,0 +1,90 @@
+
+namespace ABC {
+
+typedef float_type transformer(const float_type &);
```
there's more than meets the eye here

pearsonca replied:
:roll_eyes:

tjhladish commented on include/AbcSmc/Parameter.h:
```diff
@@ -0,0 +1,107 @@
+            // these are the core methods that must be implemented by a Parameter
+
+            // if this is an integral parameter, flatten it to the appropriate integer, then recast to double
+            // overriding this should be managed by mixin `TParameter`
```
no longer true?

pearsonca commented on include/AbcSmc/TypeDefs.h:
this should probably just go away - had to add it when dealing with some weird type issues

tjhladish commented on src/AbcSmc.cpp:
```diff
@@ -37,279 +40,408 @@ bool file_exists(const char *fileName) {
-bool AbcSmc::parse_config(string conf_filename) {
-    if (not file_exists(conf_filename.c_str())) {
-        cerr << "File does not exist: " << conf_filename << endl;
+Json::Value prepare(const string & configfile) {
```
because "file" can reasonably be a filename or a filehandle, I prefer to be explicit about this

tjhladish commented on src/AbcSmc.cpp:
```diff
@@ -390,8 +527,8 @@ bool AbcSmc::process_database(const gsl_rng* RNG) {
 
-            if (use_transformed_pars) {
-                vector<double> upars = do_complicated_untransformations(_model_pars, noised_pars.row(i));
+            if (_par_modification_map.size()) {
```
Add comment explaining what this conditional means

tjhladish commented on src/AbcSmc.cpp:
```diff
@@ -675,7 +814,7 @@ bool AbcSmc::build_database(const gsl_rng* RNG) {
         _db_execute_stringstream(db, ss);
 
-        if (use_transformed_pars) {
+        if (_par_modification_map.size()) {
```
ibid

tjhladish commented on src/AbcSmc.cpp:
```diff
@@ -715,8 +849,8 @@ bool AbcSmc::build_database(const gsl_rng* RNG) {
 
-        if (use_transformed_pars) {
-            vector<double> upars = do_complicated_untransformations(_model_pars, parrow);
+        if (_par_modification_map.size()) {
```
blarg.  some other way to make these a bit more readable, rather than commenting all of them?  the flag was easier to make sense of regarding intent

pearsonca replied:
perhaps. propose: issue it for now, and i'll get rid of this problem when i do the DB revisions.

tjhladish commented on tests/Makefile:
```diff
@@ -4,7 +4,7 @@ default: test_all
 CPP := g++
 
-CFLAGS := -O2 -Wall -std=c++17 --pedantic
+CFLAGS := -O2 -Wall -std=c++20 --pedantic
```
Should be 17?

tjhladish reviewed (changes requested):
Please take a look at line items

pearsonca commented:
have addressed the comments aside from documenting the `if (_par_modification_map ...` bits - proposed solution is to leave as is for now, create an issue, and eliminate these ifs as part of planned change to approach for DB.

Rough idea there is that these ifs will just become roughly `for(auto upar : _par_modification_map)` => insert into upar table. that reads pretty clearly to me "for each untransformed par, insert its value into the upar table".
